ddr/ddr/Models/Leveling/Spawner.swift

```diff
@@ -0,0 +1,68 @@
+
+import Foundation
+import SpriteKit
+
+protocol Spawnable: Hashable {
+  func despawn()
+}
+
+class Spawner<T: Spawnable>{
+  
+  private let maxSpawned: Int
+  private let minSpawned: Int
+  private let maxConcurrent: Int
+  private let cooldown: TimeInterval
+  private let pSpawn: CGFloat
+  private let getNewSpawn: (Spawner) -> T
+  
+  private var canSapwnAfter: Date
+  
+  private var currentlySpawned: Set<T>
+  private var totalSpawned: Int
+  
+  
+  
+  init(maxSpawned: Int, minSpawned: Int, maxConcurrent: Int, cooldown: TimeInterval, expectedDuration: TimeInterval, getNewSpawn: @escaping (Spawner) -> T) throws {
+    self.maxSpawned = maxSpawned
+    self.minSpawned = minSpawned
+    self.maxConcurrent = maxConcurrent
+    self.getNewSpawn = getNewSpawn
+    self.cooldown = cooldown
+    
+    self.canSapwnAfter = Date()
+    self.currentlySpawned = []
+    self.totalSpawned = 0
+    
+    if self.minSpawned > 0 {
+      self.pSpawn = CGFloat(self.minSpawned) * (CGFloat(GameScene.FPS) / CGFloat(expectedDuration))
```

Comment:
Might add some comments to show what this block is doing? I'm not sure what pSpawn is

Reply:
pSpawn is the probability that the spawner should spawn when spaw is called. It is set using the expected duration and assuming that spawn will be called at the framerate of the game scene (i.e. in the update loop)

I added some comments